topology/probes/libvirt/libvirt.go

```diff
@@ -0,0 +1,309 @@
+}
+
+// DomainStateMap stringifies the state of a domain
+var DomainStateMap map[libvirtgo.DomainState]string = map[libvirtgo.DomainState]string{
+	libvirtgo.DOMAIN_NOSTATE:     "Undefined",
+	libvirtgo.DOMAIN_RUNNING:     "UP",
+	libvirtgo.DOMAIN_BLOCKED:     "BLOCKED",
+	libvirtgo.DOMAIN_PAUSED:      "PAUSED",
+	libvirtgo.DOMAIN_SHUTDOWN:    "DOWN",
+	libvirtgo.DOMAIN_CRASHED:     "CRASHED",
+	libvirtgo.DOMAIN_PMSUSPENDED: "PMSUSPENDED",
+	libvirtgo.DOMAIN_SHUTOFF:     "DOWN",
+}
+
+// Interface is XML coding of an interface in libvirt
+type Interface struct {
+	Mac struct {
+		Address string `xml:"address,attr"`
+	} `xml:"mac"`
+	Target struct {
+		Device string `xml:"dev,attr"`
+	} `xml:"target"`
+	Address Address `xml:"address"`
+	Alias   struct {
+		Name string `xml:"name,attr"`
+	} `xml:"alias"`
+	Host *graph.Node `xml:"-"`
+}
+
+// Domain is the subset of XML coding of a domain in libvirt
+type Domain struct {
+	Interfaces []Interface `xml:"devices>interface"`
+}
+
+// queryDomain uses libvirt to get information on the interfaces of a domain.
+// domain is the libvirt domain queried and bag is an accumulator for results
+func (probe *Probe) queryDomain(
+	domain *libvirtgo.Domain, // domain to query
+	domainNode *graph.Node, // Node representing the domain
+	bag *[]*Interface, // to accumulate results
+	constraint string, // to restrict the search to a single interface (by alias)
+) {
+	rawXML, err := domain.GetXMLDesc(0)
+	if err != nil {
+		logging.GetLogger().Errorf("Cannot get XMLDesc: %s", err)
+		return
+	}
+	d := Domain{}
+	if err = xml.Unmarshal([]byte(rawXML), &d); err != nil {
+		logging.GetLogger().Errorf("XML parsing error: %s", err)
+		return
+	}
+	for i := range d.Interfaces {
+		itf := &d.Interfaces[i]
+		if constraint == "" || constraint == itf.Alias.Name {
+			itf.Host = domainNode
+			*bag = append(*bag, itf)
+		}
+	}
+}
+
+// registerInterfaces puts the information collected in the graph
+// interfaces is an array of collected information.
+func (probe *Probe) registerInterfaces(interfaces []*Interface) {
+	for _, itf := range interfaces {
+		name := itf.Target.Device
+		if name == "" {
+			continue
+		}
+		logging.GetLogger().Debugf(
+			"Libvirt interface %s on %s", name, itf.Host)
+		probe.tunProcessor.DoAction(itf, name)
+	}
+}
+
+// ProcessNode adds the libvirt interface information to a node of the graph
+func (itf *Interface) ProcessNode(g *graph.Graph, node *graph.Node) bool {
+	logging.GetLogger().Debugf("enrich %s", itf.Alias.Name)
+	tr := g.StartMetadataTransaction(node)
+	tr.AddMetadata("Libvirt.MAC", itf.Mac.Address)
+	tr.AddMetadata("Libvirt.Domain", itf.Host)
+	address := itf.Address
+	formatted := fmt.Sprintf(
+		"%s:%s.%s.%s.%s", address.Type, address.Domain, address.Bus,
+		address.Slot, address.Function)
+	tr.AddMetadata("Libvirt.Address", formatted)
+	tr.AddMetadata("Libvirt.Alias", itf.Alias.Name)
+	tr.AddMetadata("PeerIntfMAC", itf.Mac.Address)
+	tr.Commit()
+	go func() {
+		g.Lock()
```

Comment:
Why taking the lock for the link and not for the metadata ?
When called from `OnNodeAdded`, ProcessNode has the lock so you don't have to care about it.
When called from `DoAction`, there is no lock. So I suggest you make sure that DoAction is executed under the graph lock so you don't have to care about it in `ProcessNode`.

Reply:
There is a real problem with the graph API : 
* you have so called transactions and they do not care taking the lock. For metadata, nodes should have their own lock. That would be much cleaner.
* When called from OnNodeAdded, you probably have a Read Lock not a Write Lock. This is at least what you have with regular event listener and promoting a read lock is not possible.
So the solution is probably to put everything under the go func()

Reply:
@pierrecregut In OnNodeAdded, you have a Write lock. So you safely update the metadata and create the link. Then, in DoAction, you can also get a Write lock so that ProcessNode is protected.

Reply:
ok. the confusion came from the fact that eventHandler have their own lock (read locked in notifyListeners) are part of the graph and are named g...